src/websockethandler.cpp

```diff
@@ -0,0 +1,178 @@
+  }
+
+  qDebug() << "Attemtping to open WebSocket connection";
+  m_reconnectOnClose = true;
+
+  QNetworkRequest request;
+  request.setRawHeader("Authorization",
+                       SettingsHolder::instance()->token().toLocal8Bit());
+  request.setUrl(QUrl(WEBSOCKET_SERVER_URL));
+  m_webSocket.open(request);
+}
+
+/**
+ * @brief Closes the websocket connection.
+ *
+ * No-op in case the connection is already closed.
+ *
+ * Closing the websocket connection through this function does keeps it closed,
+ * unless the connection is explicitly reopened through
+ * `WebSocketHandler::open`.
+ */
+void WebSocketHandler::close() {
+  if (m_webSocket.state() == QAbstractSocket::UnconnectedState ||
+      m_webSocket.state() == QAbstractSocket::ClosingState) {
+    qDebug() << "Attempted to close a WebSocket, but it's already closed.";
+    return;
+  }
+
+  qDebug() << "Closing WebSocket";
+  m_reconnectOnClose = false;
+
+  m_webSocket.close();
+}
+
+/**
+ * @brief Sends a ping to the WebSocket server.
+ *
+ * WebSocket connections may silently be broken. Periodically pinging the server
+ * keeps track of the connection. If verified that the connection has been
+ * broken the connection is properly closed, which triggers new connection
+ * attempts.
+ */
+void WebSocketHandler::ping() {
+  if (!this->m_isAlive) {
+    qDebug() << "WebSocket was disconnected silently. Closing.";
+    m_webSocket.close();
+    return;
+  }
+
+  m_isAlive = false;
+  m_webSocket.ping();
+}
+
+/**
+ * @brief Acknowledges a pong response from the WebSocket server.
+ *
+ * @param elapsedTime How long it took to get the response since a ping was sent
+ * out.
+ */
+void WebSocketHandler::onPong(quint64 elapsedTime) {
+  qDebug() << "WebSocket pong" << elapsedTime;
+  m_isAlive = true;
+}
+
+/**
+ * @brief Ackowledges the WebSocket has been succesfully connected.
+ */
+void WebSocketHandler::onConnected() {
+  qDebug() << "WebSocket connected";
+
+  m_isAlive = true;
+  connect(&m_webSocket, &QWebSocket::textMessageReceived, this,
+          &WebSocketHandler::onMessageReceived);
+
+  // Start ping timer.
+  m_pingTimer = new QTimer(this);
+  connect(m_pingTimer, &QTimer::timeout, this, &WebSocketHandler::ping);
+  m_pingTimer->start(WEBSOCKET_PING_INTERVAL);
+}
+
+/**
+ * @brief Ackowledges the WebSocket has been closed.
+ *
+ * May trigger a reconnection attempt depending on the `m_reconnectOnClose`
+ * state.
+ */
+void WebSocketHandler::onClose() {
+  m_pingTimer->stop();
+
+  if (m_reconnectOnClose) {
+    qDebug() << "WebSocket closed. Attempting to reopen in:"
+             << WEBSOCKET_RETRY_INTERVAL;
+
+    QTimer::singleShot(WEBSOCKET_RETRY_INTERVAL, this, &WebSocketHandler::open);
+  } else {
+    qDebug() << "WebSocket closed explicitly. Will not attempt to reopen.";
+  }
+}
+
+/**
+ * @brief Ackowledges there was a WebSocket error.
+ *
+ */
+void WebSocketHandler::onError(QAbstractSocket::SocketError error) {
+  qDebug() << "WebSocket error:" << error;
```

Comment:
here i would try to reconnect (after X seconds).

Reply:
You can do it using:

`TimerSingleShot::create(this, X-seconds, [this]() { reconnect(); });`

Reply:
If the Websocket is closed, the onClose handler is also called and that already attempts a reconnect.

Reply:
Ack, I just realized I am confusing this. On Node.js everytime there is an error it is also followed by a close event. It seems it's not really the case on Qt. I'll fix this.

Reply:
Addressed on [4f882d5](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/3505/commits/4f882d5bef590f04c0cf0be2c8f6e6f537f1c280)